### PR TITLE
Enforce code style on doc comments #156

### PR DESCRIPTION
**Pull request by alankritdabral**
Summary: 
Added format_rust_code.sh file that uses rustfmt tool on rust code in fences.
to check document formatting run: run bash format_rust_code.sh --mode check
to fix run : run bash format_rust_code.sh --mode fix

modifie qc.yml: To run format_rust_code.sh

Issue References:
Fixes: #156 
/claim #156 

Screenshot from 2023-12-17 13-58-39

### CODE REVIEW

**Comment by alankritdabral** [comment]
@koraa can you tell me if this approach is correct

**Comment by alankritdabral** [comment]
@koraa

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,44 @@
+name: Run Formater
```
Comment (ologbonowiwi) [inline .github/workflows/ci.yml:1]: Can you change this file name to `lint.yml`? It'd make more sense as has no continuous integration related stuff on it.

**Review by ologbonowiwi — CHANGES_REQUESTED** [review]
IMO it also needs to be checked  if the project will actually use prettier or if it'll use other tooling to handle it. 

I know you took inspiration from tailcallhq/tailcall, but every project has different needs and we could have checked with @koraa if this is the desired approach _before_ working on the implementation.

**Comment by alankritdabral** [comment]
> /tip 10

Thank you for  the tip. Should i close the pr as @Min2who wants to get it assigned or should i keep working on it

**Comment by koraa** [comment]
> > /tip 10
> 
> Thank you for the tip. Should i close the pr as @Min2who wants to get it assigned or should i keep working on it

You are invited to keep working on this or collaborate with @ologbonowiwi or @Min2who or both and if the bounty is low or all three of you want to share in the bounty, I can raise it a bit too (in this case you would have to post a bid in this issue); I am sure either could help you figure out how to tackle this issue.

I'd rather enable people to work on issues than get bogged down by assigning ownership :)

The basic two steps to (1) format examples in the readme and (2) format examples in doc comments can be tackled in parallel anyway!

**Comment by koraa** [comment]
@alankritdabral Do you intend to finish this project?

**Comment by alankritdabral** [comment]
> @alankritdabral Do you intend to finish this project?
Well i have one approach in mind will be pushing some commits tomorrow. Sorry for the delay 🙏🏻😔😔.

**Comment by alankritdabral** [comment]
@koraa can you take a look in the code changes you can even run locally

File: format_rust_code.sh [hunk 2]
```diff
@@ -0,0 +1,42 @@
+#!/bin/bash
```
Comment (wucke13) [inline format_rust_code.sh:1]: Please use `#!/usr/bin/env bash`.

**Comment by wucke13** [comment]
The approach doesn't yet work (but it can be fixed!).

1. If there is a formatting error, please make sure that the CI job fails. Best case you set an internal variable, but keep checking all files, and only after all files have been checked return the 0 if no error was found or 1 if at least one file violates the formatting.
2. Some rust doc code directly starts with `let var = ...`. This is detected as a syntax error. You have to enclose this into a `fn my_fn(){ }` or something similar to make sure that you don't get syntax errors from `rustfmt`.

**Comment by alankritdabral** [comment]
@wucke13 can you run the workflow and review it 
if want to run it locally you can run bash format_rust_code.sh --mode fix

**Comment by alankritdabral** [comment]
@wucke13 🤓

**Comment by koraa** [comment]
Hi @alankritdabral,

this is pretty hacky…but also pretty cool. It only solves half of the issue – i.e. formatting examples in markdown code while disregarding markdown code in rustdoc but be that as it may, I am happy to merge this and award the bounty for just that improvement.

If I run `--fix` and then `--check`, check still reports issues, can you fix that?

> @wucke13 what to do with non-utf8 codes rustfmt cant deal with it should i just ignore non-utf8 ??

Can you elaborate on what you mean by non-utf8? All our source files should be utf8-encoded.

**Comment by alankritdabral** [comment]
> If I run `--fix` and then `--check`, check still reports issues, can you fix that?

i have fixed this issue will be pushing in a new pr

**Comment by alankritdabral** [comment]
> Can you elaborate on what you mean by non-utf8? All our source files should be utf8-encoded.

well it was a mistake from my side thee  code was treating  anything inside fence  code with `\` as a escape sequence so it was giving an error as non valid utf code i am almost done with this issue

**Comment by alankritdabral** [comment]
@koraa i have tested the file now its working fine

**Comment by alankritdabral** [comment]
@koraa

**Comment by wucke13** [comment]
@alankritdabral I took the liberty to squash your commits and apply the tool as well. Thank you for your contribution!

**Comment by alankritdabral** [comment]
> @alankritdabral I took the liberty to squash your commits and apply the tool as well. Thank you for your contribution!

just wanted to know if there is any community for the rosenpass for eg ddiscord?

**Review by wucke13 — APPROVED** [review]

**Comment by koraa** [comment]
> > @alankritdabral I took the liberty to squash your commits and apply the tool as well. Thank you for your contribution!
> 
> just wanted to know if there is any community for the rosenpass for eg ddiscord?

Yes there is, you would be most welcome to join our matrix developer chat :)

https://matrix.to/#/#rosenpass-dev:matrix.org

**Pull request merged**